libraries/tickfs/SPEC.md

```diff
@@ -0,0 +1,509 @@
+region 2, then try again in region 0.
+
+### Adding a second key
+
+When a new key TWO with a `[u8; 32]` `value` is added with `append_key()` the
+flash might look like this:
+
+```
+0x000                  0x400                  0x800                 0xC00
+--------------------------------------------------------------------------
+|||||     Region 0     |||||     Region 1     |||||     Region 2     |||||
+|||||                  |||||                  |||||                  |||||
+|||||                  |||||ONE|TWO           |||||                  |||||
+--------------------------------------------------------------------------
+```
+
+Where TWO will have the same structure as ONE, except with a different hash
+value, different checksum and starts at address 0x54C. Note that depending
+on the hash of TWO it could be added to any region.
+
+### Adding a third key
+
+When a new key THRID with a `[u8; 32]` `value` is added with `append_key()` the
+flash might look like this
+
+```
+0x000                  0x400                  0x800                 0xC00
+--------------------------------------------------------------------------
+|||||     Region 0     |||||     Region 1     |||||     Region 2     |||||
+|||||                  |||||                  |||||                  |||||
+|||||THRID             |||||ONE|TWO           |||||                  |||||
+--------------------------------------------------------------------------
+```
+
+Note that although region 1 isn't full, it is placed in region 0 based on the
+region calculated from the lower byte of it's hash.
+
+Objects will be placed in chronological order inside a region. Objects will
+not be placed in order inside flash though. Instead they are placed at a
+region depending on their hash. This is described in more detail above.
+
+### Finding keys
+
+Locating keys in flash follows a similar process to writing them.
+
+First we calculate the hash and determine the region where we expect the
+object to be stored. With the same key input and hash function we will always
+get the same hash and region.
+
+Using the same example as above, we will get this:
+
+```
+Key:    "ONE"
+Hash:   0xeda10078886193bb
+Region: 1 (0x93bb % 3)
+```
+
+Then we load the entire region from flash. In this example that would be
+region 1. So we read the entire region 1 from flash.
+
+We then iterate over the loaded region, starting with the first byte.
+
+We check to make sure the version is supported and that the object isn't
+marked as !`valid`.
+
+We then check to see if the object hash matches the hash we are
+looking for. If it doesn't we move forward in the loaded region by the
+total length of the object we just checked and start the process again.
+
+We continue this loop until we either find the key we are looking for or
+find a version 0xFF, indicating the end of the blocks in that region.
+
+This method allows a quick retrieval of data from a given key. We only
```

Comment:
When all regions are full then yes. A lookup will search all of flash.

I can't really think of a way to avoid doing that though. Having some sort of list of keys would use space, and I think space is going to be too important for that.